### DISCUSSION TITLE
What are we going to do about `nightly-version`

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#2905] [Q&A]
In `rust-bitcoin` we have a github action that runs twice a week updating the Rust nightly toolchain version in `nightly-version`. As we start to use the `bitcoin-maintainer-tools` repo for CI in a bunch of other crates we are getting one of these files in each repo but in no other repo are we updating the file.

Merging the toochain update requires:
- checkout the PR
- install the new toolchain
- ack the PR
- checkout master
- run the github-merge script to merge the PR

This process leaves the old nightly version installed taking up a non-trivial amount of storage (over a gig IIRC last time my disk filled up).

## Questions

1. How are we going to handle updating the proliferation of `nightly-version` files in each repo?
2. Can we centralize (dirty word I know) the version file and update process somehow?
3. At the very least we should have a script in `bitcoin-maintainer-tools` that does all the steps listed above (for merging the toolchain update change)

**Comment by apoelstra** [comment]
We can try "centralizing" it in `bitcoin-maintainer-tools` but we will need to manually/cronjob update the pin for `bitcoin-maintainer-tools`.

Probably what we want to do is cronjob-update `nightly-version` on the first of the month or something, in everything except `rust-bitcoin` where it's OK to keep doing it more frequently, since this is the most active repo and it'll give us a heads up about upcoming changes.

I have multiple free terabytes on all the systems I develop on and am not too put out by proliferating rust-nightly versions. If it's a problem for you I think we should reduce the frequency on rust-bitcoin to weekly or even bi-weekly.

**Reply by apoelstra** [reply]
Whatever we do, we should make sure that the dates we update on rust-bitcoin are a superset of the dates we update everything else. So the rust-bitcoin nightly versions are always usable for the other crates.

**Reply by tcharding** [reply]
Updating all the other repos to whatever `rust-bitcoin` was on the 1st of each month sounds like a good idea. We could almost just do that manually for a month or two to see how it goes.

**Comment by nyonson** [comment]
This is a tad tangential, but I stumbled on this question when trying to figure out a good way to stop messing up my local nightly fmt toolchain. Has the [rust-toolchain.toml](https://rust-lang.github.io/rustup/overrides.html#the-toolchain-file) override setting been considered? I believe it might simplify some of the CI setup (just impliclty use this instead of cat'ing the nightly-version file) as well as make it easier on devs if the tooling is auto-updated for them.

**Reply by Kixunil** [reply]
The toolchain file doesn't take into account that we pin different versions for different tasks - MSRV because it's MSRV, stable for API checking and nightly for formatting and linting. [Ideally the version for formatting should be in `rustfmt.toml`](https://rust-lang.github.io/rustfmt/?version=v1.6.0&search=#required_version)

**Reply by apoelstra** [reply]
I wouldn't mind moving the "home" of the nightly version to rustfmt.toml much like the "home" of the MSRV version is clippy.toml. But this would greatly complicate command-line parsing since there isn't very good support for TOML in bash.

FWIW it's rarely rustfmt that causes breakage for us on nightly (though with our current cronjob-formatting setup, we might not notice breakage..). It's usually clippy, and less often than that, rustc itself.

**Reply by Kixunil** [reply]
I believe the point was to make sure contributors format with the same version. Which to me makes enough sense to try at least hack it: just `sed` the version in toml to the current one with each update of the pin in addition to what we do currently. I think the string is unique enough for this to work fine.
